### PR TITLE
Make tests run depending on available backends

### PR DESCRIPTION
**Pull request by notoraptor**
Hi! This is a PR to prepare tests refactoring so that they could run with only available backends.

Now tests should run only with available backends, ie., for e.g., if relay backend is available and pytorch is not availabe, then tests should run only with relay.

However, I don't know how to check it with contiguous integration.

At least, now, I want to check if current tests will pass. We could also use this PR to discuss how we must re-factorize the tests. What do you think ? @abergeron @breuleux

### CODE REVIEW

**Comment by notoraptor** [comment]
@breuleux tests passed ! I just updated the coverage settings.

**Comment by notoraptor** [comment]
Hi @breuleux @abergeron ! I am rebasing and updating this PR, and I have activated testing for all available backends in many test files, including `test_grad`. But, now, I get an error with `relay` backend for test `test_list_while` in `tes_grad`: https://github.com/notoraptor/myia/blob/module-tests/tests/test_grad.py#L380 

Function `default_convert` in `ressource` module receives a `tvm.runtime.ADT` object and is then unable to convert it: https://github.com/notoraptor/myia/blob/module-tests/myia/pipeline/resources.py#L57

It seems the ADT object is generated by relay backend from a tagged value and then passed to myia VM which is unable to handle it. I don't yet know how to fix this bug.

I am also waiting for travis and jenkins here to check if there are some other issues.

```
(myia) notoraptor@notoraptor-linux:~/mila/dev/git/myia$ pytest -xvvs tests/test_grad.py::test_list_while
=========================================================================================== test session starts ===========================================================================================
platform linux -- Python 3.7.9, pytest-5.4.3, py-1.9.0, pluggy-0.13.1 -- /home/notoraptor/anaconda3/envs/myia/bin/python
cachedir: .pytest_cache
rootdir: /media/win/Users/notoraptor/mila/dev/git/myia, inifile: pytest.ini
plugins: cov-2.8.1
collected 2 items                                                                                                                                                                                         

tests/test_grad.py::test_list_while[pytorch-cpu-grad] <- myia/testing/multitest.py In:  in0=[1.0, 2.0, 3.0, 4.0]
Out: test_list_while=24.0
dtest_list_while/din0.0 OK: == 24.0
dtest_list_while/din0.1 OK: == 12.0
dtest_list_while/din0.2 OK: == 8.0
dtest_list_while/din0.3 OK: == 6.0
PASSED
tests/test_grad.py::test_list_while[relay-cpu-grad] <- myia/testing/multitest.py FAILED

================================================================================================ FAILURES =================================================================================================
_____________________________________________________________________________________ test_list_while[relay-cpu-grad] _____________________________________________________________________________________

self = <myia.pipeline.resources.ConverterResource object at 0x7f1832137d10>
value = ((<tvm.nd.NDArray shape=(), cpu(0)>
array(1.), runtime.ADT(0x55c924cf29a0)), <tvm.nd.NDArray shape=(), cpu(0)>
array(1.)), manage = True

    def __call__(self, value, manage=True):
        """Convert a value."""
        if isinstance(value, Graph) and value.abstract is not None:
            return value
    
        try:
>           v = self.object_map[value]
E           KeyError: ((<tvm.nd.NDArray shape=(), cpu(0)>
E           array(1.), runtime.ADT(0x55c924cf29a0)), <tvm.nd.NDArray shape=(), cpu(0)>
E           array(1.))

myia/pipeline/resources.py:108: KeyError

During handling of the above exception, another exception occurred:

self = <myia.pipeline.resources.ConverterResource object at 0x7f1832137d10>, value = (<tvm.nd.NDArray shape=(), cpu(0)>
array(1.), runtime.ADT(0x55c924cf29a0)), manage = True

    def __call__(self, value, manage=True):
        """Convert a value."""
        if isinstance(value, Graph) and value.abstract is not None:
            return value
    
        try:
>           v = self.object_map[value]
E           KeyError: (<tvm.nd.NDArray shape=(), cpu(0)>
E           array(1.), runtime.ADT(0x55c924cf29a0))

myia/pipeline/resources.py:108: KeyError

During handling of the above exception, another exception occurred:

self = <myia.pipeline.resources.ConverterResource object at 0x7f1832137d10>, value = runtime.ADT(0x55c924cf29a0), manage = True

    def __call__(self, value, manage=True):
        """Convert a value."""
        if isinstance(value, Graph) and value.abstract is not None:
            return value
    
        try:
>           v = self.object_map[value]
E           KeyError: runtime.ADT(0x55c924cf29a0)

myia/pipeline/resources.py:108: KeyError

During handling of the above exception, another exception occurred:

test = <myia.testing.multitest.MyiaFunctionTest object at 0x7f1843797f90>

    def runtest(test):
>       test.run(fn)

myia/testing/multitest.py:77: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
myia/testing/multitest.py:176: in run
    return self.runtest(self, fn, **self.spec)
tests/test_grad.py:195: in gradient
    fn, fn, args, pipeline=pipeline, rel_error=rel_error, argspec=abstract
tests/test_grad.py:159: in _grad_test
    gtest.assert_match()
myia/debug/finite_diff.py:298: in assert_match
    results = self.compare()
myia/debug/finite_diff.py:279: in compare
    exact = self.compute_exact()
myia/debug/finite_diff.py:223: in compute_exact
    grads = self.gfn(*self.clean_args, self.unwrap(out_sen))
myia/pipeline/steps.py:518: in wrapped
    res = fn(*args)
myia/vm.py:71: in __call__
    return self.vm.evaluate(self.graph, args, closure=self.values)
myia/vm.py:172: in evaluate
    args = self.convert(tuple(_args))
myia/pipeline/resources.py:113: in __call__
    v = default_convert(self, value)
/home/notoraptor/anaconda3/envs/myia/lib/python3.7/site-packages/ovld/core.py:195: in default_convert.dispatch
    return method(*args, **kwargs)
myia/pipeline/resources.py:45: in default_convert[*, tuple]
    return type(seq)(env(x) for x in seq)
myia/pipeline/resources.py:45: in <genexpr>
    return type(seq)(env(x) for x in seq)
myia/pipeline/resources.py:113: in __call__
    v = default_convert(self, value)
/home/notoraptor/anaconda3/envs/myia/lib/python3.7/site-packages/ovld/core.py:195: in default_convert.dispatch
    return method(*args, **kwargs)
myia/pipeline/resources.py:45: in default_convert[*, tuple]
    return type(seq)(env(x) for x in seq)
myia/pipeline/resources.py:45: in <genexpr>
    return type(seq)(env(x) for x in seq)
myia/pipeline/resources.py:113: in __call__
    v = default_convert(self, value)
/home/notoraptor/anaconda3/envs/myia/lib/python3.7/site-packages/ovld/core.py:195: in default_convert.dispatch
    return method(*args, **kwargs)
myia/pipeline/resources.py:59: in default_convert[*, *]
    if hasattr(x, "__to_myia__"):
/home/notoraptor/anaconda3/envs/myia/lib/python3.7/site-packages/tvm/runtime/object.py:56: in __getattr__
    return _ffi_node_api.NodeGetAttr(self, name)
tvm/_ffi/_cython/./packed_func.pxi:321: in tvm._ffi._cy3.core.PackedFuncBase.__call__
    ???
tvm/_ffi/_cython/./packed_func.pxi:256: in tvm._ffi._cy3.core.FuncCall
    ???
tvm/_ffi/_cython/./packed_func.pxi:245: in tvm._ffi._cy3.core.FuncCall3
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

>   ???
E   TypeError: Traceback (most recent call last):
E     [bt] (5) /home/notoraptor/anaconda3/envs/myia/lib/libtvm.so(TVMFuncCall+0x65) [0x7f1835e2f745]
E     [bt] (4) /home/notoraptor/anaconda3/envs/myia/lib/libtvm.so(+0x6962d4) [0x7f183550d2d4]
E     [bt] (3) /home/notoraptor/anaconda3/envs/myia/lib/libtvm.so(+0x6983c2) [0x7f183550f3c2]
E     [bt] (2) /home/notoraptor/anaconda3/envs/myia/lib/libtvm.so(tvm::ReflectionVTable::GetAttr(tvm::runtime::Object*, tvm::runtime::String const&) const+0x1b1) [0x7f183550eae1]
E     [bt] (1) /home/notoraptor/anaconda3/envs/myia/lib/libtvm.so(+0x697903) [0x7f183550e903]
E     [bt] (0) /home/notoraptor/anaconda3/envs/myia/lib/libtvm.so(+0x69748f) [0x7f183550e48f]
E     File "/home/user/conda-bld/tvm-libs_1596554067552/work/include/tvm/node/reflection.h", line 390
E   TypeError: runtime.ADT is not registered via TVM_REGISTER_NODE_TYPE

tvm/_ffi/_cython/./base.pxi:160: TypeError
========================================================================================= short test summary info =========================================================================================
FAILED tests/test_grad.py::test_list_while[relay-cpu-grad] - TypeError: Traceback (most recent call last):
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
======================================================================================= 1 failed, 1 passed in 6.49s =======================================================================================
```

**Comment by notoraptor** [comment]
Hi @breuleux @abergeron Ultimately, it seems `test_list_while` is the only one remaining failing test, so I just skipped it for realy backend (I added a way to run "all backends except given ones").

I checked the relay pattern matching, but it seems it detects the match at compile time ( https://tvm.apache.org/docs/langref/relay_pattern.html ). I did not make further research to check if it can wait for runtime before running.

Currently, all tests pass, but now I have a strange sphinx doc error.

**Comment by breuleux** [comment]
@notoraptor The sphinx doc error is likely from the `__doc__` that `ovld` generates automatically. Try adding a docstring to the first function in `to_abstract_test`, if that doesn't work we might need to merge #375 first because I solved some at least one of the doc issues in `ovld 0.2.x` but because of breaking changes these versions require a few adjustments that are in #375 (this PR is limited to `ovld 0.1.9`).

**Comment by notoraptor** [comment]
@breuleux I added a docstring, but it does not seem to fix the problem. It's strange, because some other uses of ovld in myia code don't generate such errors.

File: tests/compile/test_backend.py [hunk 1]
```diff
@@ -1,42 +1,15 @@
...
-        assert parse_default() == ("pytorch", {"target": "cpu"})
-
-        os.environ["MYIA_BACKEND"] = "relay?target=cpu&device_id=0"
-        assert parse_default() == ("relay", {"target": "cpu", "device_id": "0"})
-    finally:
-        # Make sure we don't switch the default for other tests.
-        if before is None:
-            del os.environ["MYIA_BACKEND"]
-        else:
-            os.environ["MYIA_BACKEND"] = before
```
Comment (abergeron) [inline tests/compile/test_backend.py:39]: I think that this should stay here.  This is a test of the string parser and doesn't actually require importing any backend.

**Review by abergeron — COMMENTED** [review]
Overall I think it's ok, but there are some areas that I'm not so sure about.

**Pull request merged**